File: two_pointers/container_with_most_water/container_with_most_water.go
// Package containswater will find the area of the greatest amount of water stored.
package containswater

func containsWater(heights []int) int {
	var result int
	left, right := 0, len(heights)-1

	for left < right {
		leftValue := heights[left]
		rightValue := heights[right]
		// Compute area, right is the width and height is the left or right value.
		area := (right - 1) * min(leftValue, rightValue)

		if area > result {
			result = area
		}
		if leftValue < rightValue {
			left += 1
			continue
		}
		right -= 1
	}

	return result
}

func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

File: two_pointers/container_with_most_water/container_with_most_water.py
def contains_water(heights):
    result = 0
    left, right = 0, len(heights) - 1
    while left < right:
        # right - left is the difference to determine the width.
        area = (right - left) * min(heights[left], heights[right])
        result = max(result, area)
        if heights[left] > heights[right]:
            right -= 1
            continue
        left += 1
    return result
